istioctl/pkg/multicluster/remote_secret.go

```diff
@@ -0,0 +1,270 @@
+// Copyright 2019 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package multicluster
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+	"k8s.io/apimachinery/pkg/runtime/serializer/versioning"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
+	_ "k8s.io/client-go/plugin/pkg/client/auth" // to avoid 'No Auth Provider found for name "gcp"'
+	"k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/client-go/tools/clientcmd/api/latest"
+
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/kube/secretcontroller"
+)
+
+var (
+	Codec  runtime.Codec
+	Scheme *runtime.Scheme
+)
+
+func init() {
+	Scheme = runtime.NewScheme()
+	utilruntime.Must(v1.AddToScheme(Scheme))
+	yamlSerializer := json.NewYAMLSerializer(json.DefaultMetaFactory, Scheme, Scheme)
+	Codec = versioning.NewDefaultingCodecForScheme(
+		Scheme,
+		yamlSerializer,
+		yamlSerializer,
+		v1.SchemeGroupVersion,
+		runtime.InternalGroupVersioner,
+	)
+}
+
+type Options struct {
+	ServiceAccountName string
+	Contexts           []string
+
+	// inherited from root command
+	Namespace  string
+	Kubeconfig string
+}
+
+var (
+	defaultSecretPrefix       = "istio-pilot-remote-secret-"
+	DefaultServiceAccountName = "istio-pilot-service-account"
+)
+
+// NewCreatePilotRemoteSecretCommand creates a new command for joining two clusters togeather in a multi-cluster mesh.
+func NewCreatePilotRemoteSecretCommand(kubeconfig, namespace *string) *cobra.Command {
+	o := Options{
+		ServiceAccountName: DefaultServiceAccountName,
+	}
+
+	c := &cobra.Command{
+		Use:   "create-pilot-remote-secrets <cluster_context>... [OPTIONS]",
+		Short: "Create secret(s) for Pilot to discovery and join remote k8s service registries",
+		Example: `
+# Create a secret for remote cluster c1 and install it in cluster c0.
+istioctl x create-pilot-remote-secrets c1 > c1-secret.kubeconfigSecretYaml
+kubectl --context=c0 -n istio-system apply -f c1-secret.kubeconfigSecretYaml
```

Comment:
I mentioned this in few other PRs: k8s does indeed allow a yaml file that contains multiple 'contexts' mixed together.
However Istio and Pilot are currently using one-secret-per-cluster: you add/remove single-context kubeconfigs to enable access to remote clusters, etc. I think it is simpler - and more consistent - to follow the same style and recommend users to maintain separate kubeconfig files for each cluster.

If you look inside, the structure of multi-context is more complicated, there are more flags needed, etc.

So I would say:
KUBECONFIG=remote_cluster.yaml istioctl x create-remote-secret > remote_cluster.secret.yaml
KUBECONFIG=primary_cluster.yaml kubectl apply -f remote_cluster.secret.yaml
rm -f remote_cluster.seret.yaml.

Reply:
Or even pipe them ( the KUBECONFIG can be used as a flag as well - I used env because this is defined by k8s, so a bit easier to remember than a istioctl-specific flag)

Reply:
And we can support both - just wanted the default/first choice to be the consistent/simpler one.